sql: improve memory usage attribution for some exec components

yuzefovich opened the pull request:
This PR contains several commits that improve memory attribution of some components to show up correctly in EXPLAIN ANALYZE and in `sql.mem.distsql.current` metric. See each commit for details.

Touches: #100548

Epic: None
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/160289)
<!-- Reviewable:end -->

yuzefovich committed:
sql: minor improvements around memory monitors

Use `FlowCtx.Mon` in one spot where available. Also simplify handling of
memory monitors with column backfiller while guaranteeing that it's
closed in an erroneous case during init.

Release note: None

yuzefovich committed:
sql: improve memory usage attribution for some exec components

Previously, we exposed the txn memory monitor as `Planner.Mon()` and it
was used in several places. In some cases, usage of the txn-bound
monitor was justified (e.g. because the allocation outlives a single
query and is txn-scoped, like a cursor), but in most cases this was done
out of convenience. A concrete short-coming of that approach was that
the memory usage of query-scoped allocations against the txn monitor do
not show up in EXPLAIN ANALYZE nor count against
`sql.mem.distsql.current` metric.

This commit fixes this misattribution by instantiating a new "exec"
monitor stored in the connExecutor that is started and stopped for each
query. (In theory, we could have avoided this start-stop dance and just
kept the monitor started all the time, but stopping it ensures that all
allocations are properly released, so it seems worth it. Additionally,
starting a monitor clears its MaximumBytes watermark which we use to
update the top-level query stats.) On the main query path it's clear
when to start and stop, but on other code paths that utilize the distsql
engine and might access the exec monitor, we used `resetPlanner` as the
guide where starting-stopping behavior is needed.

This new exec monitor is a child of the txn monitor but is the parent of
the flow monitor. (Remote flows remain unaffected.) Thus, we now use the
exec monitor for EXPLAIN ANALYZE and for the distsql metric, which
allows us to capture most of the usage that was previously against the
txn monitor.

Also due to an existing issue with seemingly-incorrect order of shutdown
things around txns and pausable portals, we introduce a separate
pausable-portal exec monitor that is started only once, akin to the txn
monitor. Whenever appropriate, this ppExec monitor is used instead of
the exec monitor.

Additionally, this commit creates a separate memory monitor for the
vectorized ordered aggregator so that its memory usage was attributed
correctly. (There is a general problem that using "streaming" memory
account or allocator attributes the memory usage to the "flow" monitor
which might not be desired. Fixing that is left as a TODO.)

Release note: None

yuzefovich committed:
sem/builtins: simplify singleDatumAggregateBase

This commit simplifies `singleDatumAggregateBase` to always assume
non-nil `SingleDatumAggMemAccount` is set on the `eval.Context`. These
things were introduced long time ago in 0ac5468dbb43f06e769e476fe46d5a50b47aad41,
and at the time - out of caution - we kept the fallback behavior in case
this field is left unset. I don't think that caution is warranted, and
we do expect this field to be always set, so this commit takes advantage
of this assumption.

Release note: None

yuzefovich committed:
sem/builtins: fix attribution of memory used by some aggregates

This commit replaces usage of the planner's monitor in favor of the
unlimited monitor that `SingleDatumAggMemoryAccount` is bound to for
several aggregate builtins. This will make it so that the memory usage
of those aggregates is properly attributed to the aggregator (for the
purposes of EXPLAIN ANALYZE).

This commit also clarifies that `SingleDatumAggMemoryAccount` must be
bound to an unlimited monitor (aggregates that touch this don't know how
to spill to disk), and it fixes the windower to respect this
requirement.

Release note: None

ZhouXing19 reviewed (commented):
Thanks for doing this! I'm not familiar with this part of code so left some newbie questions. I haven't finished reviewing all the usages of TxnMon and ExecMon but wanted to give some comments I have so far.

> @ZhouXing19 partially reviewed 30 files and made 6 comments.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/160289#-:-OkHTgrn9CEruNhM_BNj:b-w5z7wg)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @yuzefovich).

___
*[`pkg/sql/sem/eval/context.go` line 207 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/160289#-OkApwg59OjWMuni0Yzo:-OkApwg59OjWMuni0Yzp:b-7pp4hu) ([raw file](https://github.com/cockroachdb/cockroach/blob/911f797cbdabc73f0d5a4637b101d8081f0515e1/pkg/sql/sem/eval/context.go#L207)):*
> ```Go
> 	//
> 	// It **must** be bound to an unlimited memory monitor.
> 	SingleDatumAggMemAccount *mon.BoundAccount
> ```

Sorry, this might be an obvious one, does this `**must** be bound to an unlimited memory monitor` only bound to `evalCtx.SingleDatumAggMemAccount`, or any other "singleDatumAggregateAccount"? 

It's because I see we have `makeSingleDatumAggregateBase()`, where `singleDatumAggregateBase.acc` is generated from `evalCtx.Planner.Mon().MakeBoundAccount()`. And it seems to me that the "unlimited" is not necessarily enforced here. 

So is it because it doesn't matter for `singleDatumAggregateBase.acc` to go unlimited or not, or the rule is implicitly enforced already? 
___
*[`pkg/sql/backfill.go` line 2959 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/160289#-Ok1dmfdCiZgNPRKqrmH:-Ok1dmfdCiZgNPRKqrmI:b3pwtj6) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b78252f25384953d96303b2cf931690707a3ea2/pkg/sql/backfill.go#L2959)):*
> ```Go
> 	}
> 	var columnBackfillerMon *mon.BytesMonitor
> 	if evalCtx.Planner.Mon() != nil {
> ```

ditto, why the nil check is removed...?
___
*[`pkg/sql/conn_executor_exec.go` line 3509 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/160289#-OkHMVzoDw_yWQzs-ixK:-OkHMVzoDw_yWQzs-ixL:b-vhskfa) ([raw file](https://github.com/cockroachdb/cockroach/blob/08fc20f90e30ee8f227763642f8e889ea0403ded/pkg/sql/conn_executor_exec.go#L3509)):*
> ```Go
> 	p := &ex.planner
> 	ex.resetPlanner(ctx, p, nil, now)
> 	ex.execMon.StartNoReserved(ctx, ex.state.txnMon)
> ```

Hmm, is it possible that when we start the execMon here, the parent txnMon hasn't started yet? It's because here we can be in no-txn state here, and the txnMon won't be started when the txn state machine steps from noTxn to open. Would this be problematic?
___
*[`pkg/sql/backfill/backfill.go` line 165 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/160289#-Ok1dcRt3nJPxXFuDhiP:-Ok1dcRu5G9--Z8Q5nkd:b-vaxey6) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b78252f25384953d96303b2cf931690707a3ea2/pkg/sql/backfill/backfill.go#L165)):*
> ```Go
> 	}
> 
> 	// Create a bound account associated with the column backfiller.
> ```

Sorry this might be an obvious one, but why do we remove these nil check? 

Update: Is it similar to the second commit that we assume they are always not nil?
___
*[`pkg/sql/conn_executor.go` line 3059 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/160289#-OkHC_wHDiOeaHl3aB3M:-OkHC_wIDyIjznFYvcnu:b2exr8b) ([raw file](https://github.com/cockroachdb/cockroach/blob/08fc20f90e30ee8f227763642f8e889ea0403ded/pkg/sql/conn_executor.go#L3059)):*
> ```Go
> 
> 	defer func() {
> 		defer ex.execMon.Stop(ctx)
> ```

1. nit: Is the defer-in-defer intentional? If so, is it because the `ex.planner.maybeLogStatement()` below also accounts for memory usage? 
2. IIUC we have connExecutor.close() accounts for stopping execMon already, but we also close execMon here. Is it because closing the monitor is idempotent? 
3. If stopping the execMon is necessary here, does it mean we need to do the same for the defer in execCopyIn?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Partial feedback is always welcome!

> @yuzefovich made 6 comments.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/160289#-:-Ok_2Y_T9SQkyalJfShQ:b-y35otr)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @ZhouXing19).

___
*[`pkg/sql/backfill.go` line 2959 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/160289#-Ok1dmfdCiZgNPRKqrmH:-Ok_2bY6BYbdrSpH_0o2:b-bvcg49) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b78252f25384953d96303b2cf931690707a3ea2/pkg/sql/backfill.go#L2959)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

ditto, why the nil check is removed...?
</blockquote></details>

Because it should actually never be nil here, and if it is, then something is wrong. This code is probably a leftover from the past when memory accounting was just added to the backfill.
___
*[`pkg/sql/conn_executor.go` line 3059 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/160289#-OkHC_wHDiOeaHl3aB3M:-Ok_Bdwa1ybbi-I4amne:b-74zfhr) ([raw file](https://github.com/cockroachdb/cockroach/blob/08fc20f90e30ee8f227763642f8e889ea0403ded/pkg/sql/conn_executor.go#L3059)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

1. nit: Is the defer-in-defer intentional? If so, is it because the `ex.planner.maybeLogStatement()` below also accounts for memory usage? 
2. IIUC we have connExecutor.close() accounts for stopping execMon already, but we also close execMon here. Is it because closing the monitor is idempotent? 
3. If stopping the execMon is necessary here, does it mean we need to do the same for the defer in execCopyIn?
</blockquote></details>

1. it is intentional in a sense that I wanted (needed?) for `Stop` to be executed at the end of the deferred cleanup function. I don't remember the details but I _think_ simply deferring right after the exec mon is started below led to some problems, so _probably_ there was some dependency on whatever is happening within this large defer.
2. yes, it's idempotent (although it doesn't say so in the comment). (We already rely on this for the txn mon which is stopped at the end of the txn and then will be stopped on the connExecutor closure.)
3. I alluded to this in the commit message for the last commit, but `Stop`ping the exec monitor is not _strictly necessary_ after each query execution - it's a nice to have since we enforce that on `Stop` all children monitors must have been stopped too, so stopping the exec mon after each query execution checks that all monitors created during the execution are properly stopped.
Also - answering your question more directly - stopping of the exec monitor is added into the cleanup function returned by `preparePlannerForCopy`
___
*[`pkg/sql/conn_executor_exec.go` line 3509 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/160289#-OkHMVzoDw_yWQzs-ixK:-Ok_316a94xrx5UhpvD7:bv0ivsj) ([raw file](https://github.com/cockroachdb/cockroach/blob/08fc20f90e30ee8f227763642f8e889ea0403ded/pkg/sql/conn_executor_exec.go#L3509)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

Hmm, is it possible that when we start the execMon here, the parent txnMon hasn't started yet? It's because here we can be in no-txn state here, and the txnMon won't be started when the txn state machine steps from noTxn to open. Would this be problematic?
</blockquote></details>

Huh, good catch. Indeed, it looks like we would start the exec mon before the txn mon is started a bit later. I switched to using `sessionMon` here which is the parent of the txn mon and is running through the whole lifetime of the connExecutor.
___
*[`pkg/sql/backfill/backfill.go` line 165 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/160289#-Ok1dcRt3nJPxXFuDhiP:-Ok_Ak9P5WfUyvdwv5Q0:b-imren0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b78252f25384953d96303b2cf931690707a3ea2/pkg/sql/backfill/backfill.go#L165)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

Sorry this might be an obvious one, but why do we remove these nil check? 

Update: Is it similar to the second commit that we assume they are always not nil?
</blockquote></details>

These were redundant and leftover from the past when the monitor could be nil, and it no longer can be.
___
*[`pkg/sql/sem/eval/context.go` line 207 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/160289#-OkApwg59OjWMuni0Yzo:-Ok_B1w2-oyAUA3knNfF:b-shmlj6) ([raw file](https://github.com/cockroachdb/cockroach/blob/911f797cbdabc73f0d5a4637b101d8081f0515e1/pkg/sql/sem/eval/context.go#L207)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

Sorry, this might be an obvious one, does this `**must** be bound to an unlimited memory monitor` only bound to `evalCtx.SingleDatumAggMemAccount`, or any other "singleDatumAggregateAccount"? 

It's because I see we have `makeSingleDatumAggregateBase()`, where `singleDatumAggregateBase.acc` is generated from `evalCtx.Planner.Mon().MakeBoundAccount()`. And it seems to me that the "unlimited" is not necessarily enforced here. 

So is it because it doesn't matter for `singleDatumAggregateBase.acc` to go unlimited or not, or the rule is implicitly enforced already?
</blockquote></details>

Both `Planner.TxnMon()` and `Planner.ExecMon()` are unlimited as in they will approve allocations up to `--max-sql-memory` node-wide. During the execution we create "limited" monitors which - in addition to the global --max-sql-memory limit - also have a "local" limit. This is how `distsql_workmem` is respected and forces some computation to spill to disk. However, not every computation / object knows how to spill to disk, so in those cases we generally expect to use "unlimited" monitor, and aggregates that use `SingleDatumAggMemAccount` are in exactly this category.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice! Just one nit suggestion/question.

> @DrewKimball reviewed 52 files and all commit messages, and made 2 comments.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/160289#-:-Ok_ylw4Dt99YN9ko6qn:bduxxl3)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich and @ZhouXing19).

___
*[`pkg/sql/flowinfra/flow.go` line 697 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/160289#-Ok_yVxmC8biZ4_TOjVa:-Ok_yVxmC8biZ4_TOjVb:b-lugnt5) ([raw file](https://github.com/cockroachdb/cockroach/blob/0209ac46d1fdc232734c6984cc4fb7058c3aee49/pkg/sql/flowinfra/flow.go#L697)):*
> ```Go
> 				FlowStats: execinfrapb.FlowStats{
> 					// Here we use ParentMon and not Mon so that the memory
> 					// allocated during logical planning is included.
> ```

nit: It seems a little sad to have to plumb ParentMon just for this purpose. Is there any way we could correct the MaxMemUsage using the parent higher up wherever it's used?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> @yuzefovich made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/160289)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball and @ZhouXing19).

___
*[`pkg/sql/flowinfra/flow.go` line 697 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/160289#-Ok_yVxmC8biZ4_TOjVa:-Oka7S7j5rywSKG9Quau:b-rxdjkh) ([raw file](https://github.com/cockroachdb/cockroach/blob/0209ac46d1fdc232734c6984cc4fb7058c3aee49/pkg/sql/flowinfra/flow.go#L697)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

nit: It seems a little sad to have to plumb ParentMon just for this purpose. Is there any way we could correct the MaxMemUsage using the parent higher up wherever it's used?
</blockquote></details>

Hm, I share your dislike of the current approach, but I don't see how it can be easily done cleaner.

We take this `ComponentStats` proto and put into the trace, so we could update it whenever we're extracting the proto from the trace. That is done in `execinfrapb.ExtractStatsFromSpans`, and plumbing the exec mon there is doable and would look roughly equal to what we have here, but then we'd additionally need to distinguish the ComponentStats proto for the local flow from those for the remote flows (since we only want to update the local one), so we'd have to extend the proto with SQL instance ID or "gateway" boolean, which together feels worse than what we currently have. Do you see a better approach?


<!-- Sent from Reviewable.io -->

ZhouXing19 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks for the detailed explanation!

> @ZhouXing19 reviewed 49 files and all commit messages, and made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/160289#-:-OkdHNwa-T_mteussbAe:b-lf5th2)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @DrewKimball).



<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> @DrewKimball made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/160289)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @yuzefovich and @ZhouXing19).

___
*[`pkg/sql/flowinfra/flow.go` line 697 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/160289#-Ok_yVxmC8biZ4_TOjVa:-OkdcXJa4XLmdcGb-I0o:buo1eo6) ([raw file](https://github.com/cockroachdb/cockroach/blob/0209ac46d1fdc232734c6984cc4fb7058c3aee49/pkg/sql/flowinfra/flow.go#L697)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Hm, I share your dislike of the current approach, but I don't see how it can be easily done cleaner.

We take this `ComponentStats` proto and put into the trace, so we could update it whenever we're extracting the proto from the trace. That is done in `execinfrapb.ExtractStatsFromSpans`, and plumbing the exec mon there is doable and would look roughly equal to what we have here, but then we'd additionally need to distinguish the ComponentStats proto for the local flow from those for the remote flows (since we only want to update the local one), so we'd have to extend the proto with SQL instance ID or "gateway" boolean, which together feels worse than what we currently have. Do you see a better approach?
</blockquote></details>

Could we update just the query level stats in populateQueryLevelStats, and leave the flow stats as they are?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> @yuzefovich made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/160289)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @DrewKimball, @kyle-a-wong, and @ZhouXing19).

___
*[`pkg/sql/flowinfra/flow.go` line 697 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/160289#-Ok_yVxmC8biZ4_TOjVa:-Oke5AhaCrurLAMz_nL8:b-4c3mok) ([raw file](https://github.com/cockroachdb/cockroach/blob/0209ac46d1fdc232734c6984cc4fb7058c3aee49/pkg/sql/flowinfra/flow.go#L697)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Could we update just the query level stats in populateQueryLevelStats, and leave the flow stats as they are?
</blockquote></details>

Hm, I don't think it's any cleaner - I just pushed a WIP commit (that is incomplete) where we remove ParentMon (so Mon - which is the flow monitor - is used for the gateway flow) and then we update the FlowStats component just for the gateway flow after the query execution. Is this what you had in mind? To me it looks more ugly than plumbing ParentMon into the flow context. WDYT?

(This exercise did expose a potential problem for inner plans where we currently would use the exec mon of the planner which is tracking the outer and the inner plans, but for the inner we probably want to have a child exec mon.)


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> @DrewKimball made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/160289)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @yuzefovich and @ZhouXing19).

___
*[`pkg/sql/flowinfra/flow.go` line 697 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/160289#-Ok_yVxmC8biZ4_TOjVa:-OkemoNtBtMIA9yGOmfR:br0vj7f) ([raw file](https://github.com/cockroachdb/cockroach/blob/0209ac46d1fdc232734c6984cc4fb7058c3aee49/pkg/sql/flowinfra/flow.go#L697)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Hm, I don't think it's any cleaner - I just pushed a WIP commit (that is incomplete) where we remove ParentMon (so Mon - which is the flow monitor - is used for the gateway flow) and then we update the FlowStats component just for the gateway flow after the query execution. Is this what you had in mind? To me it looks more ugly than plumbing ParentMon into the flow context. WDYT?

(This exercise did expose a potential problem for inner plans where we currently would use the exec mon of the planner which is tracking the outer and the inner plans, but for the inner we probably want to have a child exec mon.)
</blockquote></details>

I'm probably wrong... but wouldn't it be enough to just modify queryLevelStats like this?
```
diff --git a/pkg/sql/conn_executor_exec.go b/pkg/sql/conn_executor_exec.go
index 772033ab3b6..4361d142b55 100644
--- a/pkg/sql/conn_executor_exec.go
+++ b/pkg/sql/conn_executor_exec.go
@@ -3081,6 +3081,7 @@ func populateQueryLevelStats(
                        // we need to zero it out.
                        ih.queryLevelStatsWithErr.Stats.AdmissionWaitTime = 0
                }
+               ih.queryLevelStatsWithErr.Stats.MaxMemUsage = p.execMon.MaximumBytes()
        }
        if ih.traceMetadata != nil && ih.explainPlan != nil {
                ih.traceMetadata.annotateExplain(
```
It seems reasonable to consider planning memory/time as being separate from the gateway flow, and I *think* this would still result in the total memory being displayed at the root of EXPLAIN ANALYZE.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTRs!

I'll wait for the green CI and will merge then.

> @yuzefovich made 2 comments.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/160289#-:-OkfB-MgCRp5Yb9d2Adk:b1k6p8v)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @DrewKimball and @ZhouXing19).

___
*[`pkg/sql/flowinfra/flow.go` line 697 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/160289#-Ok_yVxmC8biZ4_TOjVa:-Okf9iyfEtIs_zodWjU0:b-95voup) ([raw file](https://github.com/cockroachdb/cockroach/blob/0209ac46d1fdc232734c6984cc4fb7058c3aee49/pkg/sql/flowinfra/flow.go#L697)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

I'm probably wrong... but wouldn't it be enough to just modify queryLevelStats like this?
```
diff --git a/pkg/sql/conn_executor_exec.go b/pkg/sql/conn_executor_exec.go
index 772033ab3b6..4361d142b55 100644
--- a/pkg/sql/conn_executor_exec.go
+++ b/pkg/sql/conn_executor_exec.go
@@ -3081,6 +3081,7 @@ func populateQueryLevelStats(
                        // we need to zero it out.
                        ih.queryLevelStatsWithErr.Stats.AdmissionWaitTime = 0
                }
+               ih.queryLevelStatsWithErr.Stats.MaxMemUsage = p.execMon.MaximumBytes()
        }
        if ih.traceMetadata != nil && ih.explainPlan != nil {
                ih.traceMetadata.annotateExplain(
```
It seems reasonable to consider planning memory/time as being separate from the gateway flow, and I *think* this would still result in the total memory being displayed at the root of EXPLAIN ANALYZE.
</blockquote></details>

Indeed, nice, not sure why I didn't see it. (I think I was confused by several locations in which we call `ExtractStatsFromSpans` which probably can be consolidated. Left a TODO for that.)


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
Thanks!

> @DrewKimball reviewed 54 files and all commit messages, and made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/160289#-:-OkfDF4g2RX25huygEMV:b-ucfknh)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @ZhouXing19).



<!-- Sent from Reviewable.io -->

yuzefovich commented:
bors r+
